### PR TITLE
Add search and support for subdirectories to icon picker

### LINKED ISSUE #2476: Add a setting to recursively search through the icons directory
### Role

Due to the amount of icons I have amassed from downloading modpacks and icons for my other instances, I started organizing them in multiple sub-directories like "Vanilla" (a 3D block model for every major vanilla Minecraft version), "Modpacks", "Mods", "Resource Packs" and "Projects"

### Suggestion

I want that in the settings, below/near the place where you set the icon directory, there be a button to toggle searching it recursively, similar to the button below the downloads directory with the same purpose

### Benefit

My files stay organized AND I can have lots of pretty pictures for my instance list

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

One problem that might be relevant is that this would still bloat the actual dialogue window where you select the icons with like hundreds of icons strewn around haphazardly. One idea to counter that is that if this setting is enabled, icons found in sub-directories are separated by a collapsible line with the sub-directories name similar to the lines present with the groups feature in the main instance list. Perhaps the built-in selection of icons could also be sorted into a "Built-in" category this way?

### PR DESCRIPTION
**Pull request by QazCetelic**
Adds subdirectory functionality for use case #2476 and search functionality

# Main changes
## Searching / filtering
- text ui element (`launcher/ui/dialogs/IconPickerDialog.cpp:38`)
- proxy model for filtering (`launcher/ui/dialogs/IconPickerDialog.cpp:40`)
- emit event to prevent incorrect indices of proxy model (`launcher/icons/IconList.cpp:470`)

## Support for subdirectories
- new `addPathRecursively` & `removePathRecursively` methods
- sort change to put all icons subdirectories last
- icon key uses relative path to icons directory instead of basename
- open folder button follows directory of selected icon (`IconList::iconDirectory`)

# Showcase
## Searching / filtering
image

## Support for subdirectories
Folder structure of test
image

It currently shows the name of the subdirectory separated by » to avoid confusion when an icon is present in two directories.
image
image

### CODE REVIEW

File: launcher/icons/IconList.cpp [hunk 1]
```diff
@@ -425,3 +494,15 @@ QString IconList::getDirectory() const
...
     return m_dir.absolutePath();
 }
+
+/// Returns the directory of the icon with the given key or the default directory if it's a builtin icon.
+QString IconList::iconDirectory(const QString& key) const
+{
+    for (const auto& mmc_icon : icons) {
+        if (mmc_icon.m_key == key && mmc_icon.has(IconType::FileBased)) {
+            QFileInfo icon_file(mmc_icon.getFilePath());
+            return icon_file.dir().path();
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]: Why not just return the absolute path of the icon(https://doc.qt.io/qt-6/qfileinfo.html#absolutePath)
```suggestion
            return icon_file.absolutePath();
```

File: launcher/icons/IconList.cpp [hunk 2]
```diff
@@ -77,10 +77,90 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+
+    // Add the directory itself
+    watching = m_watcher->addPath(path);
+
+    // Add all subdirectories
+    QFileInfoList entries = dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot);
+    for (const QFileInfo& entry : entries) {
+        if (addPathRecursively(entry.absoluteFilePath())) {
+            watching = true;
+        }
```
Comment (Trial97) [inline launcher/icons/IconList.cpp:106]: You can write the if as single or operation(not sure if this is more readable or not):
```suggestion
        watching |= addPathRecursively(entry.absoluteFilePath());
```

File: launcher/icons/IconList.cpp [hunk 3]
```diff
@@ -77,10 +77,90 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+void IconList::removePathRecursively(const QString& path)
+{
+    QFileInfo file_info(path);
+    if (file_info.isFile()) {
+        // Remove the icon belonging to the file
+        QString key = m_dir.relativeFilePath(file_info.absoluteFilePath());
+        int idx = getIconIndex(key);
+        if (idx == -1)
+            return;
+
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]: This code only checks the existence of the icon, it doesn't remove it.
So either this is not intended(you need to add the remove code) or this code is not needed.

File: launcher/icons/IconList.cpp [hunk 4]
```diff
@@ -77,10 +77,90 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+        QString first = directories.takeFirst();
+        QDir dir(first);
+        for (QString& file_name : dir.entryList(QDir::AllDirs | QDir::Files | QDir::NoDotAndDotDot, QDir::Name)) {
+            QString full_path = dir.filePath(file_name);  // Convert to full path
+            QFileInfo file_info(full_path);
+            if (file_info.isDir())
+                directories.push_back(full_path);
+            else
+                icon_files.push_back(full_path);
+        }
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
       for (auto fileInfo : dir.entryInfoList(QDir::AllDirs | QDir::Files | QDir::NoDotAndDotDot, QDir::Name)) {
        if (fileInfo.isDir()) {
            directories.push_back(fileInfo.absoluteFilePath());
        } else {
            icon_files.push_back(fileInfo.absoluteFilePath());
        }
    }
```

File: launcher/icons/IconList.cpp [hunk 5]
```diff
@@ -77,10 +77,90 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+        // Remove all files within the directory
+        for (const QFileInfo& file : dir.entryInfoList(QDir::Dirs | QDir::NoDotAndDotDot)) {
+            removePathRecursively(file.absoluteFilePath());
+        }
+    }
+}
+
+QStringList IconList::getIconFilePaths() const
+{
+    QStringList icon_files{};
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]: Sugestion use camelCase for variable names

File: launcher/icons/IconList.cpp [hunk 6]
```diff
@@ -77,10 +77,90 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+    return icon_files;
+}
+
+QString formatName(const QDir& icons_dir, const QFileInfo& file)
+{
+    if (file.dir() == icons_dir)
+        return file.baseName();
+
+    constexpr auto delimiter = " » ";
+    QString relative_path_without_extension = icons_dir.relativeFilePath(file.dir().path()) + QDir::separator() + file.baseName();
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]: Sugestion use camelCase for variable names

**Comment by QazCetelic** [comment]
Thanks for the feedback, I will change it soon

**Comment by QazCetelic** [comment]
I just noticed a regression which causes an issue the first time a file is renamed while being in a subdirectory after intially loading. I have marked it as draft until I find the cause.

File: launcher/icons/IconList.cpp [hunk 7]
```diff
@@ -77,10 +77,86 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+        if (addPathRecursively(entry.absoluteFilePath())) {
+            watching = true;
+        }
+    }
+    return watching;
+}
+
+void IconList::removePathRecursively(const QString& path)
+{
+    QFileInfo file_info(path);
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
    QFileInfo fileInfo(path);
```

File: launcher/icons/IconList.cpp [hunk 8]
```diff
@@ -77,10 +77,86 @@ IconList::IconList(const QStringList& builtinPaths, QString path, QObject* paren
...
+            if (fileInfo.isDir())
+                directories.push_back(fileInfo.absoluteFilePath());
+            else
+                iconFiles.push_back(fileInfo.absoluteFilePath());
+        }
+    }
+    return iconFiles;
+}
+
+QString formatName(const QDir& icons_dir, const QFileInfo& file)
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
QString formatName(const QDir& iconsDir, const QFileInfo& file)
```

File: launcher/icons/IconList.cpp [hunk 9]
```diff
@@ -95,15 +171,11 @@ void IconList::directoryChanged(const QString& path)
         if (!FS::ensureFolderPathExists(m_dir.absolutePath()))
             return;
     m_dir.refresh();
-    auto new_list = m_dir.entryList(QDir::Files, QDir::Name);
-    for (auto it = new_list.begin(); it != new_list.end(); it++) {
-        QString& foo = (*it);
-        foo = m_dir.filePath(foo);
-    }
+    QStringList new_file_names_list = getIconFilePaths();
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
    QStringList newFileNamesList = getIconFilePaths();
```

File: launcher/icons/IconList.cpp [hunk 10]
```diff
@@ -123,21 +195,16 @@ void IconList::directoryChanged(const QString& path)
...
+    for (const auto& remove : to_remove) {
         qDebug() << "Removing " << remove;
-        QFileInfo rmfile(remove);
-        QString key = rmfile.completeBaseName();
-
-        QString suffix = rmfile.suffix();
-        // The icon doesnt have a suffix, but it can have other .s in the name, so we account for those as well
-        if (!IconUtils::isIconSuffix(suffix))
-            key = rmfile.fileName();
+        QFileInfo removed_file(remove);
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
        QFileInfo removedFile(remove);
```

File: launcher/icons/IconList.cpp [hunk 11]
```diff
@@ -123,21 +195,16 @@ void IconList::directoryChanged(const QString& path)
     QSet<QString> to_add = new_set;
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
    QSet<QString> toAdd = newSet;
```

File: launcher/icons/IconList.cpp [hunk 12]
```diff
@@ -409,7 +474,7 @@ QIcon IconList::getIcon(const QString& key) const
 
     if (icon_index != -1)
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
    if (iconIndex != -1)
```

File: launcher/icons/IconList.cpp [hunk 13]
```diff
@@ -425,3 +490,15 @@ QString IconList::getDirectory() const
 {
     return m_dir.absolutePath();
 }
+
+/// Returns the directory of the icon with the given key or the default directory if it's a builtin icon.
+QString IconList::iconDirectory(const QString& key) const
+{
+    for (const auto& mmc_icon : icons) {
```
Comment (Trial97) [inline launcher/icons/IconList.cpp]:
```suggestion
    for (const auto& mmcIcon : icons) {
```

**Review by Trial97 — COMMENTED** [review]
Try to rename all variables(in the files you modified/around your changes at least) using camelCase.
Exception from this rule are the member variables that should be camelCase but with `m_` prefix to denote that they are members.
This is not a big deal but it would be nice.

**Comment by Trial97** [comment]
new_file_names_list has a issue on qt5

**Comment by QazCetelic** [comment]
> new_file_names_list has a issue on qt5

It's because of the preprocessing for the set initialization. I've split it into a separate function.

**Comment by QazCetelic** [comment]
This has been marked draft until now because there was a bug that caused the name to be invalid when renaming an icon file in a nested folder while the application was running. I didn't have time to take a proper look, but I had some time today and was able to fix it.

I'd love to get feedback, my experience with Qt is limited.

File: launcher/icons/IconList.cpp (hunk 2 above)
Comment (QazCetelic) [reply launcher/icons/IconList.cpp:106]: It doesn't look clearer imo

File: launcher/icons/IconList.h [hunk 14]
```diff
@@ -91,18 +92,20 @@ class IconList : public QAbstractListModel {
...
     void directoryChanged(const QString& path);
 
    protected slots:
     void fileChanged(const QString& path);
-    void SettingChanged(const Setting& setting, QVariant value);
+    void SettingChanged(const Setting& setting, const QVariant& value);
 
    private:
     shared_qobject_ptr<QFileSystemWatcher> m_watcher;
-    bool is_watching;
```
Comment (Trial97) [inline launcher/icons/IconList.h:104]: see https://github.com/PrismLauncher/PrismLauncher/blob/develop/CONTRIBUTING.md

**Review by Trial97 — APPROVED** [review]

**Review by Ryex — APPROVED** [review]

**Pull request merged**